app/src/main/java/org/wikipedia/util/ShareUtil.kt

```diff
@@ -0,0 +1,221 @@
+import android.content.pm.LabeledIntent
+import android.content.pm.ResolveInfo
+import android.graphics.Bitmap
+import android.net.Uri
+import android.os.Build
+import android.os.Parcelable
+import android.widget.Toast
+import androidx.core.content.FileProvider
+import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
+import io.reactivex.rxjava3.core.Observable
+import io.reactivex.rxjava3.schedulers.Schedulers
+import org.apache.commons.lang3.StringUtils
+import org.wikipedia.BuildConfig
+import org.wikipedia.R
+import org.wikipedia.page.PageTitle
+import org.wikipedia.util.DateUtil.getFeedCardDateString
+import org.wikipedia.util.StringUtil.fromHtml
+import org.wikipedia.util.log.L.e
+import java.io.File
+import java.util.*
+
+object ShareUtil {
+    private const val APP_PACKAGE_REGEX = "org\\.wikipedia.*"
+    private const val FILE_PROVIDER_AUTHORITY = BuildConfig.APPLICATION_ID + ".fileprovider"
+    private const val FILE_PREFIX = "file://"
+
+    private fun shareText(context: Context, subject: String, text: String) {
+        val shareIntent = Intent(Intent.ACTION_SEND)
+        shareIntent.putExtra(Intent.EXTRA_SUBJECT, subject)
+        shareIntent.putExtra(Intent.EXTRA_TEXT, text)
+        shareIntent.type = "text/plain"
+        val chooserIntent = Intent.createChooser(shareIntent, context.getString(R.string.share_via))
+        if (chooserIntent == null) {
+            showUnresolvableIntentMessage(context)
+        } else {
+            context.startActivity(chooserIntent)
+        }
+    }
+
+    @JvmStatic
+    fun shareText(context: Context, title: PageTitle) {
+        shareText(context, fromHtml(title.displayText).toString(),
+                UriUtil.getUrlWithProvenance(context, title, R.string.prov_share_link))
+    }
+
+    fun shareText(context: Context, title: PageTitle, newId: Long, oldId: Long) {
+        shareText(context, fromHtml(title.displayText).toString(),
+                title.getWebApiUrl("diff=" + newId + "&oldid=" +
+                        oldId + "&variant=" + title.wikiSite.languageCode()))
+    }
+
+    @JvmStatic
+    @SuppressLint("CheckResult")
+    fun shareImage(context: Context, bmp: Bitmap,
+                   imageFileName: String, subject: String?,
+                   text: String?) {
+        Observable.fromCallable {
+            getUriFromFile(context,
+                    processBitmapForSharing(context, bmp, imageFileName))
+        }
+                .subscribeOn(Schedulers.io())
+                .observeOn(AndroidSchedulers.mainThread())
+                .subscribe({ uri: Uri? ->
+                    if (uri == null) {
+                        displayShareErrorMessage(context)
+                        return@subscribe
+                    }
+                    val chooserIntent = buildImageShareChooserIntent(context, subject, text, uri)
+                    context.startActivity(chooserIntent)
+                }) { caught: Throwable -> displayOnCatchMessage(caught, context) }
```

Comment:
Would it be possible to have this code well-formatted?

Reply:
Lmk if it is better now?

Reply:
It is still the same. Should be something like this:
```
        Observable.fromCallable {
                    getUriFromFile(context, processBitmapForSharing(context, bmp, imageFileName))
                }
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
```

or even reduce to be one line:

```
Observable.fromCallable { getUriFromFile(context, processBitmapForSharing(context, bmp, imageFileName))}
                .subscribeOn(Schedulers.io())
```

Reply:
now?

Reply:
Looks better. Thanks.